Content.Server/Botany/Systems/PlantAnalyzerSystem.cs

```diff
@@ -0,0 +1,261 @@
+        SubscribeLocalEvent<PlantAnalyzerComponent, PlantAnalyzerSetMode>(OnModeSelected);
+    }
+
+    private void OnAfterInteract(Entity<PlantAnalyzerComponent> plantAnalyzer, ref AfterInteractEvent args)
+    {
+        if (args.Target == null || !args.CanReach || !HasComp<SeedComponent>(args.Target) && !HasComp<PlantHolderComponent>(args.Target) || !_cell.HasActivatableCharge(plantAnalyzer, user: args.User))
+            return;
+
+        _audio.PlayPvs(plantAnalyzer.Comp.ScanningBeginSound, plantAnalyzer);
+
+        var cancellationTokenSource = new CancellationTokenSource();
+        var cancellationTokenWrapper = new CancellationTokenWrapper(cancellationTokenSource);
+
+        _cancellationTokenSources[plantAnalyzer] = cancellationTokenWrapper;
+
+        if (plantAnalyzer.Comp.AdvancedScan)
+        {
+            _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, plantAnalyzer.Comp.AdvScanDelay,
+                new PlantAnalyzerDoAfterEvent { CancellationTokenWrapper = cancellationTokenWrapper }, plantAnalyzer, target: args.Target, used: plantAnalyzer)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                NeedHand = true,
+            });
+        }
+        else
+        {
+            _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, plantAnalyzer.Comp.ScanDelay,
+                new PlantAnalyzerDoAfterEvent { CancellationTokenWrapper = cancellationTokenWrapper }, plantAnalyzer, target: args.Target, used: plantAnalyzer)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                NeedHand = true
+            });
+        }
+    }
+
+    private void OnDoAfter(Entity<PlantAnalyzerComponent> component, ref PlantAnalyzerDoAfterEvent args)
+    {
+        if (component.Comp.AdvancedScan) // Double charge use for advanced scan.
+        {
+            if (!_cell.TryUseActivatableCharge(component, user: args.User))
+                return;
+        }
+        if (args.Handled || args.Cancelled || args.Args.Target == null || !_cell.TryUseActivatableCharge(component.Owner, user: args.User))
+            return;
+
+        _audio.PlayPvs(component.Comp.ScanningEndSound, args.Args.User);
+
+        OpenUserInterface(args.User, component);
+        UpdateScannedUser(component, args.Args.User, args.Args.Target.Value, component);
+
+        args.Handled = true;
+        _cancellationTokenSources.Remove(component);
+    }
+
+    private void OpenUserInterface(EntityUid user, EntityUid analyzer)
+    {
+        if (!TryComp<ActorComponent>(user, out var actor) || !_uiSystem.TryGetUi(analyzer, PlantAnalyzerUiKey.Key, out var ui))
+            return;
+
+        _uiSystem.OpenUi(ui, actor.PlayerSession);
+    }
+
+    public void UpdateScannedUser(EntityUid uid, EntityUid user, EntityUid target, Entity<PlantAnalyzerComponent> component)
```

Comment:
Isn't `uid` and `Entity<PlantAnalyzerComponent>` the same entity?

Reply:
It is... thanks for noticing.